### PR TITLE
`Communication`: Add profile picture support

### PR DESCRIPTION
**Pull request by PaRangger**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently, the readability of the communication module is lacking in terms of being able to distinguish who wrote a message. This is due to the only visual que being the username, which are hard to differentiate from another without close inspection.

### Description

I added support for profile pictures in the communication module. This was done by adding a possibility to upload a picture in the user settings. The uploaded picture will be displayed when you write a message or answer in a thread. As a default picture, a user will be assigned a random color hue, which will be taken as the background of the default picture, and their initials will be displayed in white bold text. For example "Tom Martin Muster" will be "TM". This change also required slight rework in the displaying of the post headers. The user role is not displayed as a colorful badge anymore, but simply put below the username to not overload the colors.

### Steps for Testing

Prerequisites:
- 1 Course with communication enabled
- 1 Student/Tutor/Instructor

1. Log in to Artemis
2. Go to the course
3. Go to the communication tab
4. Switch to any channel the user can write in
5. Post a new message
6. Verify the displaying of the default picture
7. On the top right, click the username and go to settings
8. Press the "Add a profile picture" button
9. Select a picture and crop accordingly
10. Revisit the communication tab and verify if the image is shown in the messages

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2
#### Performance Tests
- [ ] Test 1
- [ ] Test 2

### Screenshots

Bildschirmfoto 2024-09-05 um 14 32 14

Bildschirmfoto 2024-09-03 um 21 36 00

Bildschirmfoto 2024-09-03 um 21 36 28

Bildschirmfoto 2024-09-03 um 21 36 37

## Summary by CodeRabbit

- **New Features**
	- Users can now upload, display, and delete their profile pictures.
	- New API endpoints for managing user profile images have been added.
	- Enhanced user interface for managing profile pictures in account settings.
	- Users can now view author profile pictures in posts.
	- Improved visual representation of user roles with icons.

- **Bug Fixes**
	- Fixed issues with displaying default profile pictures when none are set.

- **Documentation**
	- Updated localization files to include error messages for profile picture uploads.

- **Tests**
	- Added tests for updating and deleting user profile pictures to ensure functionality works as expected.

### CODE REVIEW

File: src/main/webapp/app/course/manage/course-update.component.ts [hunk 1]
```diff
@@ -639,7 +639,7 @@ export class CourseUpdateComponent implements OnInit {
 
     openCropper(): void {
         const modalRef = this.modalService.open(ImageCropperModalComponent, { size: 'm' });
-        modalRef.componentInstance.courseImageUploadFile = this.courseImageUploadFile;
+        modalRef.componentInstance.uploadFile = this.courseImageUploadFile;
```
[the remark this answers is not in the document]
Comment (PaRangger) [reply src/main/webapp/app/course/manage/course-update.component.ts:642]: Checked all uses, could not find any other place where it should be renamed

**Comment by anian03** [comment]
Just tested it and it worked as expected :)
One question: Is a gravatar integration possible/planned? I think it would be nice if we could make the user's gravatar picture the default (instead of none), just like Slack or Zulip do

**Review by anian03 — CHANGES_REQUESTED** [review]
Two suggestions:
- I think it would be nice if you could click on your own profile picture in a conversation to have an option to change it there. Users might not know to go to settings to do that.
- And should we maybe also display the profile picture here?

**Comment by bassner** [comment]
Gravatar should definitely be implemented, ig. in a follow-up

**Comment by PaRangger** [comment]
> Just tested it and it worked as expected :) One question: Is a gravatar integration possible/planned? I think it would be nice if we could make the user's gravatar picture the default (instead of none), just like Slack or Zulip do

Good idea! I think we can add this in a standalone PR afterwards.

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java [hunk 2]
```diff
@@ -115,6 +126,55 @@ public ResponseEntity<Void> updateUserNotificationDate() {
...
 
+    /**
+     * PUT users/profile-picture : upload a profile picture
+     *
+     * @param file the image file that is being uploaded
+     * @return the ResponseEntity with status 200 (OK) and with body of current user
+     */
+    @PutMapping("users/profile-picture")
+    @EnforceAtLeastStudent
+    public ResponseEntity<UserDTO> updateProfilePicture(@RequestPart MultipartFile file) throws URISyntaxException {
```
Comment (SimonEntholzer) [inline src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:137]: I guess it's not entirely clear where the actual difference is, but most other settings use the `account/....` endpoint. I think also doing that for the profile picture, might align a bit better with the rest of the settings endpoints.

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]

**Comment by PaRangger** [comment]
> Two suggestions:
> 
> * I think it would be nice if you could click on your own profile picture in a conversation to have an option to change it there. Users might not know to go to settings to do that.
> * And should we maybe also display the profile picture here? 

I added the picture to the profile and added an option to navigate to settings by clicking the user icon of your own messages. 😊

Bildschirmfoto 2024-09-06 um 08 32 53
 
Bildschirmfoto 2024-09-06 um 09 30 49

File: src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java (hunk 2 above)
Comment (PaRangger) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java:137]: Hey! Yeah, you are right, in my mind the account api was used for mainly managing credentials (username, password, tokens) and the user api was used for metadata. I was a bit on the fence when deciding which one to go for. The decision was mainly influenced by the administration api also being called "admin/users". But I'll change it to /account, I think that makes more sense. 👍🏼

Comment (PaRangger) [reply src/main/java/de/tum/in/www1/artemis/web/rest/UserResource.java]: I changed the api route to /api/account/profile-picture and moved the functions to the AccountResource in this [commit](https://github.com/ls1intum/Artemis/pull/9277/commits/d9a77eb44ab7f0ade78c98d3fbb56c58b9c85ad5). 👍🏼

**Review by asliayk — APPROVED** [review]
great work :) tested locally and everything works as expected

**Review by dfuchss — COMMENTED** [review]
Please ensure that the max size of the images is very low. Keep in mind that multiple 1000 Students are registered to artemis. Also the images shall not remain if a user gets deleted (-> GDPR)

**Comment by PaRangger** [comment]
> Please ensure that the max size of the images is very low. Keep in mind that multiple 1000 Students are registered to artemis. Also the images shall not remain if a user gets deleted (-> GDPR)

Good call! I added functionality to remove the image on user deletion and adressed the image size with a max upload size restriction as well as a change of upload format to jpeg (when using the cropper). The cropper itself already is configured to scale the images to a max of 200x200. These changes should guarantee smaller filesizes in general, as well as prevent massive filesizes if anyone uploads an image to the api directly, without using the client.

**Review by anian03 — COMMENTED** [review]
Thanks for the changes! Works as expected :)

One more thing I noticed: Sometimes I get this weird black square appearing at the top left corner after/while using the photo cropper. Especially switching between light and dark mode causes this. Is this related to this PR or a general issue with the Photo Cropper?

**Comment by PaRangger** [comment]
> Thanks for the changes! Works as expected :)
> 
> One more thing I noticed: Sometimes I get this weird black square appearing at the top left corner after/while using the photo cropper. Especially switching between light and dark mode causes this. Is this related to this PR or a general issue with the Photo Cropper? 

I noticed a similar issue one time but just discarded it. I think it‘s related to the canvas that gets created by the cropper. I don‘t think its related to the changes in the PR, but I have an idea to fix it. As soon as I manage to reproduce it consistently I can create another PR to address this.

**Review by SimonEntholzer — APPROVED** [review]
Code lgtm

**Review by edkaya — APPROVED** [review]
Tested on TS6, works as expected. Really nice feature btw!

**Review by anian03 — APPROVED** [review]

**Review by BBesrour — APPROVED** [review]
Tested on ts6, lgtm!

**Review by sarpsahinalp — APPROVED** [review]
Code lgtm

File: src/main/webapp/app/shared/metis/posting-header/posting-header.directive.ts [hunk 3]
```diff
@@ -63,21 +79,83 @@ export abstract class PostingHeaderDirective<T extends Posting> implements OnIni
...
             this.userRoleBadge = roleBadgeTranslationPath + this.userAuthority;
             this.userAuthorityTooltip = toolTipTranslationPath + this.userAuthority;
         }
     }
 
+    /**
+     * Returns a pseudo-random numeric value for a given string using a simple hash function.
+     * @param {string} str - The string used for the hash function.
+     */
+    private deterministicRandomValueFromString(str: string): number {
```
[the remark this answers is not in the document]
Comment (PaRangger) [reply src/main/webapp/app/shared/metis/posting-header/posting-header.directive.ts:106]: Hi, thanks for the feedback!

Sanitization like filename transformation and hiding the path were already implemented by the File Service. I heavily oriented myself by the implementation of uploading course images. I mainly focused on validation of the file in this pr 👍

And concerning the hash function: I did not find a proper pseudo random hash function on the client side. That's why I chose to implement a very simple one for this component to generate a deterministic color hue. The hashCode function is on the server-side, or did I overlook something?

**Pull request merged**